CORE-3046 Quotas: implement deprecated notice for quota configs

pgellert opened the pull request:
This adds a warning log message to remind operators that the client quotas are now node-wide and that they are going to be deprecated in v25.1.

Fixes https://redpandadata.atlassian.net/browse/CORE-3046

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.1.x
- [ ] v23.3.x
- [ ] v23.2.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

BenPope reviewed (commented):
The implementation doesn't seem to implement the requirement:
* Create a warning log message that we show on cluster version update

Introducing new configuration in order to deprecate existing configuration seems counterproductive.

I think it is sufficient to add the log on upgrade or startup, rather than nag every hour.

michael-redpanda commented on src/v/kafka/CMakeLists.txt:
```diff
@@ -48,6 +48,7 @@ v_cc_library(
     server/protocol_utils.cc
     server/quota_manager.cc
     server/client_quota_translator.cc
+    server/monitor_deprecated_client_quota_configs.cc
```
Why not build this into `quota_manager`?  Why do we need a separate service for monitoring?

pgellert replied:
I've integrated this into `client_quota_translator` now

Deflaimun commented on src/v/config/configuration.cc:
```diff
@@ -3243,7 +3243,15 @@ configuration::configuration()
       {.needs_restart = needs_restart::yes, .visibility = visibility::user},
-      false) {}
+      false)
+  , deprecated_quota_configs_warning_interval_sec(
+      *this,
+      "deprecated_quota_configs_warning_interval_sec",
+      "Interval, in seconds, of how often to emit a message informing the "
+      "operator that the kafka client quotas cluster configs used are going to "
+      "be deprecated.",
+      {.needs_restart = needs_restart::no, .visibility = visibility::user},
+      1h) {}
```
Why does the property name starts with "deprecated"? Shouldn't we set the visibility to deprecated instead?

pgellert replied:
I've reworked this now so we no longer need this config.

pgellert commented:
After talking to @BenPope offline, I've reworked this to only trigger on startup and whenever the cluster configs are updated (and moved the code into `client_quota_translator`). This removes the need for the additional cluster config and the background fibre for checking for periodic logging.

michael-redpanda reviewed (approved):
lgtm

pgellert commented on src/v/kafka/server/client_quota_translator.cc:
```diff
@@ -292,4 +301,30 @@ client_quota_translator::get_default_config(client_quota_type qt) const {
+        || !_target_fetch_tp_rate_per_client_group().empty();
+
+    if (emit_nag) {
+        vlog(
```
Changed

pgellert commented:
[Force-pushed](https://github.com/redpanda-data/redpanda/compare/9c1222f7f772527e1f072ba88fa9ca15a3bccaa8..a8b5550040f01274c3a880574631ed52c5a4da02) to log as one line.

BenPope reviewed (commented):
Can we deprecate the configs like this, or does that break something?
```diff
diff --git a/src/v/config/configuration.cc b/src/v/config/configuration.cc
index cec8a5a407..741ca4f9c7 100644
--- a/src/v/config/configuration.cc
+++ b/src/v/config/configuration.cc
@@ -499,20 +499,22 @@ configuration::configuration()
       "Target request size quota byte rate (bytes per second) - 2GB default",
       {.needs_restart = needs_restart::no,
        .example = "1073741824",
-       .visibility = visibility::user},
+       .visibility = visibility::deprecated},
       target_produce_quota_byte_rate_default,
       {.min = 1_MiB})
   , target_fetch_quota_byte_rate(
       *this,
       "target_fetch_quota_byte_rate",
       "Target fetch size quota byte rate (bytes per second) - disabled default",
-      {.needs_restart = needs_restart::no, .visibility = visibility::user},
+      {.needs_restart = needs_restart::no,
+       .visibility = visibility::deprecated},
       std::nullopt)
   , kafka_admin_topic_api_rate(
       *this,
       "kafka_admin_topic_api_rate",
       "Target quota rate (partition mutations per default_window_sec)",
-      {.needs_restart = needs_restart::no, .visibility = visibility::user},
+      {.needs_restart = needs_restart::no,
+       .visibility = visibility::deprecated},
       std::nullopt,
       {.min = 1})
   , cluster_id(
@@ -1487,7 +1489,7 @@ configuration::configuration()
       {.needs_restart = needs_restart::no,
        .example
        = R"([{'group_name': 'first_group','clients_prefix': 'group_1','quota': 10240}])",
-       .visibility = visibility::user},
+       .visibility = visibility::deprecated},
       {},
       validate_client_groups_byte_rate_quota)
   , kafka_client_group_fetch_byte_rate_quota(
@@ -1499,7 +1501,7 @@ configuration::configuration()
       {.needs_restart = needs_restart::no,
        .example
        = R"([{'group_name': 'first_group','clients_prefix': 'group_1','quota': 10240}])",
-       .visibility = visibility::user},
+       .visibility = visibility::deprecated},
       {},
       validate_client_groups_byte_rate_quota)
   , kafka_rpc_server_tcp_recv_buf(
```

pgellert commented:
> Can we deprecate the configs like this, or does that break something?

That is the diff we will do in 2 major versions to remove the configs. Once you do that, `rpk cluster config get/set` on those cluster configs would no longer work, so `visibility::deprecated` is meant for cluster configs that "used to exist".
